enhancements/builds/volume-mounted-resources.md

```diff
@@ -0,0 +1,297 @@
+
+If a build volume's mount path conflicts with any of the paths specified above, the build volume mount will override the default.
+
+Proposed api/structs:
+Note:  DockerBuildStrategy will be updated in the same way.
+
+```go
+// SourceBuildStrategy defines input parameters specific to an Source build.
+type SourceBuildStrategy struct {
+  // From is reference to an DockerImage, ImageStream, ImageStreamTag, or ImageStreamImage from which
+  // the docker image should be pulled
+  From kapi.ObjectReference
+
+  // PullSecret is the name of a Secret that would be used for setting up
+  // the authentication for pulling the Docker images from the private Docker
+  // registries
+  PullSecret *kapi.LocalObjectReference
+
+  // Env contains additional environment variables you want to pass into a builder container.
+  Env []kapi.EnvVar
+
+  // Scripts is the location of Source scripts
+  Scripts string
+
+  // Incremental flag forces the Source build to do incremental builds if true.
+  Incremental *bool
+
+  // ForcePull describes if the builder should pull the images from registry prior to building.
+  ForcePull bool
+
+  // Volumes is a list of volumes that can be mounted by the build
+  // Only a subset of Kubernetes Volume sources are supported by builds.
+  // More info: https://kubernetes.io/docs/concepts/storage/volumes
+  Volumes []BuildVolume
+
+  // VolumeMounts are volumes to mount into the build
+  VolumeMounts []VolumeMount
+}
+```
+
+```go
+// BuildVolume describes a volume that is made available to build pods, such that it can be mounted into the build execution environment.
+// Only a subset of Kubernetes Volume sources are supported by builds.
+type BuildVolume struct {
+
+  // Volume's name.
+	// Must be a DNS_LABEL and unique within the pod.
+	// More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names
+	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+
+	// BuildVolumeSource represents the location and type of the mounted volume.
+	BuildVolumeSource `json:",inline" protobuf:"bytes,2,opt,name=volumeSource"`
+}
+
+
+// Represents the source of a volume to mount.
+// Only one of its members may be specified.
+// This is a subset of the core Kubernetes VolumeSource.
+type BuildVolumeSource struct {
+	// Secret represents a secret that should populate this volume.
+	// More info: https://kubernetes.io/docs/concepts/storage/volumes#secret
+	// +optional
+	Secret *kapi.SecretVolumeSource `json:"secret,omitempty" protobuf:"bytes,1,opt,name=secret"`
+	
+	// ConfigMap represents a configMap that should populate this volume
+	// +optional
+	ConfigMap *kapi.ConfigMapVolumeSource `json:"configMap,omitempty" protobuf:"bytes,2,opt,name=configMap"`
+
+	// CSI (Container Storage Interface) represents ephemeral storage that is handled by certain external CSI drivers (Beta feature).
+	// +optional
+	CSI *kapi.CSIVolumeSource `json:"csi,omitempty" protobuf:"bytes,3,opt,name=csi"`
```

Comment:
How will this be controlled with respect to SCCs allowed for users?  I'm actually unsure today how CSI volumes are controlled via SCC

Reply:
Due to concerns with SCCs, CSI volume sources were dropped from this EP. The build controller uses the privileged SCC to do "safe" host path mounts in build pods, and even with builds in user namespaces we will likely keep these host path mounts.

CSI volume sources can be reintroduced in the future. I think we can gate the allowed CSI driver for these volume types in one of two ways (perhaps not mutually exclusive):

- Enhance SCCs to allowlist drivers for `CSI` volume sources.
- Enhance the cluster Build controller config to allowlist drivers for CSI volume mounts.

Reply:
so if CSI volumes are being dropped from the enhancement, how does this move us towards the shared entitlement cred use case?

(don't get me wrong, it's still a useful evolution of the api and configmap/secret volume mounting has existing use cases/demand)

Reply:
We don't get the benefit of shared entitlements, but we remove developer pain with the experience of adding secrets. Especially for the Docker strategy, where developers are on their own to remove the secret and squash layers.

Adding CSI volumes need to be added in a follow-up EP, where we will need to carefully consider how we secure builds with CSI volumes.

Reply:
(note - we still get the foundation of an API to add volumes. From a coding perspective this will be 80% of what we need to do to support the shared entitlement use case).

Reply:
got it.